core/env/replace_spec.rb

```diff
@@ -1,15 +1,52 @@
 require_relative '../../spec_helper'
 
 describe "ENV.replace" do
+  before :each do
+    @orig = ENV.to_hash
+    ENV.delete("foo")
+  end
+
+  after :each do
+    ENV.replace(@orig)
+  end
 
   it "replaces ENV with a Hash" do
-    ENV["foo"] = "bar"
-    e = ENV.reject { |k, v| k == "foo" }
-    e["baz"] = "bam"
-    ENV.replace e
-    ENV["foo"].should == nil
-    ENV["baz"].should == "bam"
-    ENV.delete "baz"
+    ENV.replace("foo" => "0", "bar" => "1").should equal(ENV)
+    ENV.size.should == 2
+    ENV["foo"].should == "0"
+    ENV["bar"].should == "1"
+  end
+
+  it "raises TypeError if the argument is not a Hash" do
+    -> { ENV.replace(Object.new) }.should raise_error(TypeError, "no implicit conversion of Object into Hash")
+    ENV.to_hash.should == @orig
+  end
+
+  it "raises TypeError if a key is not a String" do
+    -> { ENV.replace(Object.new => "0") }.should raise_error(TypeError, "no implicit conversion of Object into String")
+    ENV.to_hash.should == @orig
+  end
+
+  it "raises TypeError if a value is not a String" do
+    -> { ENV.replace("foo" => Object.new).should raise_error(TypeError, "no implicit conversion of Object into String") }
```

Comment:
There were still a few issues here, I fixed them in 13a40d29a3a66f1b0dd8a8b6167d2b2d49f630a5.